app/utils/practice-system.ts

```diff
@@ -335,3 +300,69 @@ export async function queryDeckPracticeEntriesCountByQueueType(
     PRACTICE_QUEUE_TYPES.map((type) => [type, aggregated[type]?.count ?? 0])
   ) as any;
 }
+
+export function queryNextPracticeEntryRandomMode(
+  deckId: number,
+  now: Date,
+  seed: number
+) {
+  const randInt = hashInt32(seed);
+  const randUniform = randInt / 2 ** 32;
+
+  // choose queueType by a fixed probability (0.85, 0.1, 0.05)
+  const randQueueType =
+    randUniform <= 0.85 ? "NEW" : randUniform <= 0.95 ? "LEARN" : "REVIEW";
+
+  const RANDOM_MODE_SCORE_ALIAS = "randomModeScore";
+
+  const query = db
+    .select({
+      ...T.practiceEntries,
+      // RANDOM(HASH(practiceAction) ^ seedInt)  (in [0, 1))
+      // +
+      // (scheduledAt bonus)                     (in [0, 0.5])
+      // +
+      // (queueType bonus)                       (2 if randQueueType)
+      [RANDOM_MODE_SCORE_ALIAS]: sql<number>`(
+        RAND(
+          CONV(
+            SUBSTRING(
+              HEX(
+                UNHEX(SHA1(${T.practiceEntries.id})) ^
+                UNHEX(SHA1(${T.practiceEntries.updatedAt})) ^
+                UNHEX(SHA1(${randInt}))
+              ),
+              1,
+              8
+            ),
+            16,
+            10
+          )
+        )
+        +
+        0.1 * LEAST(5, (UNIX_TIMESTAMP(${now}) - UNIX_TIMESTAMP(scheduledAt)) / (60 * 60 * 24 * 7))
```

Comment:
This `scheduledAt` bonus was actually broken before, but now it might prioritize old one too much.